lib/token.cpp

```diff
@@ -2546,6 +2546,18 @@ bool Token::hasKnownSymbolicValue(const Token* tok) const
     });
 }
 
+const ValueFlow::Value* Token::getKnownValue() const
+{
+    if (!mImpl->mValues)
+        return nullptr;
+    if (mImpl->mValues->empty())
```

Comment:
this code is technically redundant.

Reply:
this redundant code is still here.
```
    if (mImpl->mValues->empty())
        return nullptr;
```
was it kept by intention?

Reply:
This:
```
    if (mImpl->mValues->empty())
        return nullptr;
    auto it = std::find_if(mImpl->mValues->begin(), mImpl->mValues->end(), [&](const ValueFlow::Value& value) {
        return value.isKnown();
    });
    return it == mImpl->mValues->end() ? nullptr : &*it;
}
```
Is logically the same if the first 2 lines are removed:
```
    auto it = std::find_if(mImpl->mValues->begin(), mImpl->mValues->end(), [&](const ValueFlow::Value& value) {
        return value.isKnown();
    });
    return it == mImpl->mValues->end() ? nullptr : &*it;
```

if `mImpl->mValues` is empty then `it` will be `mImpl->mValues->end()` and that means nullptr is returned.

So I suggest to remove the extra code.

Reply:
Now I get it. 

That would not be the same since the lamdba might still be created even though it is never used. I think I saw a similar optimization somewhere recently (but it is possible that was copying data in the capture). Need to have a look at the generated code and/or profile it.